### ISSUE TITLE
HLS stream would panic

### ISSUE THREAD

**Issue by kane50613** [serenity-rs/songbird#257]
**Songbird version:** 0.4.3

**Rust version (`rustc -V`):** rustc 1.81.0 (eeb90cda1 2024-09-04)

**Description:**

The program panicked in rusty pool thread when using `HlsRequest`

```
thread 'rusty_pool_1_thread_1' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/stream_lib-0.5.2/src/hls/mod.rs:99:9:
there is no reactor running, must be called from the context of a Tokio 1.x runtime
```

**Steps to reproduce:**

```rust
let input = HlsRequest::new(...).into();
driver.play_only_input(input);
```

**Comment by kane50613** [comment]
I guess its because `stream_lib::download_hls` tries to spawn a task, maybe should make `HlsRequest` create async instead?

**Comment by Erk-** [comment]
I think this is more of a documentation issue as making a function async means something specific which is not the case here.

**Comment by WuuBoLin** [comment]
@Erk- In this [discussion](https://github.com/serenity-rs/songbird/pull/242#discussion_r1667719208), you mentioned that `DownloadStream` does not need async to be created. While this might be true, from my understanding, `mixer` uses `create` and `create_async` in the `Compose` trait to determine if an `Input` should live on a tokio executor or a thread pool.

However, you [spawned a tokio task](https://github.com/Erk-/rsget/blob/342e78a27dd2fca5da99d07ad083406cdbc2bd52/stream_lib/src/hls/mod.rs#L99) to download HLS chunks while the HLS Input will always run without tokio runtime/context. This will fail 100% of the time if we don't wrap it with another input type and implement `create_async` or by spawning the tasks directly using Runtime handler.

In conclusion, there are two possible solutions:
1. Patch all `tokio::spawn` or `tokio::task::spawn` with `Handle::current().spawn()`  in stream_lib and update the crate version in songbird
2. Create the stream in `create_async` even though the creation itself doesn't require async

**Comment by Erk-** [comment]
> In conclusion, there are two possible solutions:
> 
> 1. Patch all `tokio::spawn` or `tokio::task::spawn` with `Handle::current().spawn()`  in stream_lib and update the crate version in songbird

I don't believe this would make a difference as `Handle::current()` panics[^1] if called outside of an executor just like `tokio::spawn` does. Though the note on `Handle::current` sounds like it would solve it in your case?

> 2. Create the stream in `create_async` even though the creation itself doesn't require async

The reason I am not a fan of this is that it signals that the function is async, but it will never yield.

[^1]: https://docs.rs/tokio/latest/tokio/runtime/struct.Handle.html#method.current

**Comment by Erk-** [comment]
Oh I see the issue now, if the task gets moved to a thread that does not have a executor in by the mixer it might fail. I am not sure why I am not running into this in my own tests?

Is the example completely fulfilling in this regard?

**Comment by kane50613** [comment]
My solution is to wrap `HlsRequest` in a new Compose implementation with `should_create_async` set to true:

```rs
use async_trait::async_trait;
use reqwest::Client;
use songbird::input::{AudioStream, AudioStreamError, Compose, HlsRequest, Input};
use symphonia_core::io::MediaSource;

pub struct SoundCloudStream(HlsRequest);

impl SoundCloudStream {
  pub fn new(client: Client, url: String) -> Self {
    Self(HlsRequest::new(client, url))
  }
}

#[async_trait]
impl Compose for SoundCloudStream {
  fn create(&mut self) -> Result<AudioStream<Box<dyn MediaSource>>, AudioStreamError> {
    Err(AudioStreamError::Unsupported)
  }

  async fn create_async(&mut self) -> Result<AudioStream<Box<dyn MediaSource>>, AudioStreamError> {
    self.0.create()
  }

  fn should_create_async(&self) -> bool {
    true
  }
}

impl From<SoundCloudStream> for Input {
  fn from(stream: SoundCloudStream) -> Self {
    Input::Lazy(Box::new(stream))
  }
}
```

**Comment by WuuBoLin** [comment]
> The reason I am not a fan of this is that it signals that the function is async, but it will never yield.

Yes i'm aware of this, and i'm not a fan of this either. It results in a colored function of a synchronized task.
However, this is currently the simplest solution without modifying the underlying mixer code.

**Comment by kane50613** [comment]
> Oh I see the issue now, if the task gets moved to a thread that does not have a executor in by the mixer it might fail. I am not sure why I am not running into this in my own tests?
> 
> Is the example completely fulfilling in this regard?

[The only usage of `HlsRequest` in this repo](https://github.com/serenity-rs/songbird/blob/fe9b156906640a8e6c8135c71d37d758958fc522/src/input/sources/ytdl.rs#L203) is wrapped within `create_async` and only uses the `create` function, so nothing would go wrong. 

Because I'm using `HlsRequest` directly and the `Compose` is marked as sync, [it will be created on a different thread](https://github.com/serenity-rs/songbird/blob/fe9b156906640a8e6c8135c71d37d758958fc522/src/driver/tasks/mixer/pool.rs#L51) in the pool which does not have access to the tokio runtime.

**Comment by Erk-** [comment]
I think #262 should fix it, you are welcome to try it out if you have some time.

**Comment by WuuBoLin** [comment]
Perfect! Thanks for your help.
